pkg/router/controller/status_test.go

```diff
@@ -1089,6 +1091,9 @@ func Test_recordIngressCondition(t *testing.T) {
 			if tc.expectChanged != changed {
 				t.Errorf("expected changed=%t, but got changed=%t", tc.expectChanged, changed)
 			}
+			if diff := cmp.Diff(baselineIngressState, original, cmpOpts...); len(diff) > 0 {
+				t.Errorf("expected original to match baseline ingress from route.Status (-want +got):\n%s", diff)
+			}
```

Comment:
> * First, it had no impact other than logging. `original` was only used for logging

Hm, I see, I was mistakenly thinking that `original` was stored in the contention tracker.  (Even though `original` is passed to`(*SimpleContentionTracker) IsChangeContended`, the method doesn't actually use the value at all.)  You're right, there should be no memory impact, so it only matters for logging.

> * Besides being confusing to modify a variable called `original`,  removing the conditions removes half the benefit of logging [here](https://github.com/openshift/router/blob/d928435a15b19261166cfa57e504df58a61a4414/pkg/router/controller/status.go#L145).

Fair enough.  It's worth keeping the conditions if it makes logging more useful.  

> * It only clears conditions on the `original` return value when the ingress has changed...that feels confusing and inconsistent.

Also fair.

Reply:
> Hm, I see, I was mistakenly thinking that `original` was stored in the contention tracker. (Even though `original` is passed to`(*SimpleContentionTracker) IsChangeContended`, the method doesn't actually use the value at all.) You're right, there should be no memory impact, so it only matters for logging.

If `IsChangeContended` doesn't use value then I'd be inclined to drop it from the function signature. It would clear up any concern I have now (or in the future) w.r.t. the shallow copy.  

I didn't really chase this through as I didn't run the tests but the diff seems small (for a future PR). 

```diff
diff --git a/pkg/router/controller/contention.go b/pkg/router/controller/contention.go
index d9e1e234..0e7df160 100644
--- a/pkg/router/controller/contention.go
+++ b/pkg/router/controller/contention.go
@@ -24,7 +24,7 @@ type ContentionTracker interface {
 	// expected state of the object at this time and may be used by the tracker to
 	// determine if the most recent update was a contention. This method does not
 	// update the state of the tracker.
-	IsChangeContended(id contentionKey, now time.Time, current *routev1.RouteIngress) bool
+	IsChangeContended(id contentionKey, now time.Time) bool
 	// Clear informs the tracker that the provided ingress state was confirmed to
 	// match the current state of this process. If a subsequent call to IsChangeContended
 	// is made within the expiration window, the object will be considered as contended.
@@ -208,7 +208,7 @@ func (t *SimpleContentionTracker) Changed(id contentionKey, current *routev1.Rou
 	}
 }
 
-func (t *SimpleContentionTracker) IsChangeContended(id contentionKey, now time.Time, current *routev1.RouteIngress) bool {
+func (t *SimpleContentionTracker) IsChangeContended(id contentionKey, now time.Time) bool {
 	t.lock.Lock()
 	defer t.lock.Unlock()
 
diff --git a/pkg/router/controller/status.go b/pkg/router/controller/status.go
index 102a5730..f06f76d8 100644
--- a/pkg/router/controller/status.go
+++ b/pkg/router/controller/status.go
@@ -177,7 +177,7 @@ func performIngressConditionUpdate(action string, lease writerlease.Lease, track
 		// value, skip updating altogether and rely on the next resync to resolve conflicts. This prevents routers
 		// with different configurations from endlessly updating the route status.
 		// TRICKY: The tracker keys off of the route UID, not the workKey.
-		if !created && tracker.IsChangeContended(contentionKey(route.UID), now, original) {
+		if !created && tracker.IsChangeContended(contentionKey(route.UID), now) {
 			log.V(4).Info("skipped update due to another process altering the route with a different ingress status value", "action", action, "workKey", workKey, "original", original)
 			return writerlease.Release, false
 		}
@@ -204,7 +204,7 @@ func performIngressConditionRemoval(action string, lease writerlease.Lease, trac
 		}
 
 		route = route.DeepCopy()
-		changed, now, latest, original := removeIngressCondition(route, routerName, condType)
+		changed, now, latest := removeIngressCondition(route, routerName, condType)
 		if !changed {
 			log.V(4).Info("no changes to route needed", "action", action, "namespace", route.Namespace, "name", route.Name)
 			// Extending the lease ensures a delay in future work ONLY for followers. Unlike in
@@ -219,8 +219,8 @@ func performIngressConditionRemoval(action string, lease writerlease.Lease, trac
 		// value, skip updating altogether and rely on the next resync to resolve conflicts. This prevents routers
 		// with different configurations from endlessly updating the route status.
 		// TRICKY: The tracker keys off of the route UID, not the workKey.
-		if tracker.IsChangeContended(contentionKey(route.UID), now, original) {
-			log.V(4).Info("skipped update due to another process altering the route with a different ingress status value", "action", action, "workKey", workKey, "original", original)
+		if tracker.IsChangeContended(contentionKey(route.UID), now) {
+			log.V(4).Info("skipped update due to another process altering the route with a different ingress status value", "action", action, "workKey", workKey, "original")
 			return writerlease.Release, false
 		}
 
@@ -328,16 +328,13 @@ func recordIngressCondition(route *routev1.Route, name, hostName string, conditi
 // removeIngressCondition removes a matching status condition if it is found.
 // It returns whether the route was changed, the time it removed the condition, and
 // a pointer to the latest and original ingress records.
-func removeIngressCondition(route *routev1.Route, name string, condType routev1.RouteIngressConditionType) (changed bool, at time.Time, latest, original *routev1.RouteIngress) {
+func removeIngressCondition(route *routev1.Route, name string, condType routev1.RouteIngressConditionType) (changed bool, at time.Time, latest *routev1.RouteIngress) {
 	for i := range route.Status.Ingress {
 		existing := &route.Status.Ingress[i]
 		if existing.RouterName != name {
 			continue
 		}
 
-		// preserve a deep copy of the original ingress
-		original := existing.DeepCopy()
-
 		at := time.Time{}
 
 		// Remove the condition if we can find it.
@@ -357,10 +354,10 @@ func removeIngressCondition(route *routev1.Route, name string, condType routev1.
 			existing.Conditions = updatedConditions
 		}
 
-		return changed, at, existing, original
+		return changed, at, existing
 	}
 
-	return false, time.Time{}, nil, nil
+	return false, time.Time{}, nil
 }
 
 // findMostRecentIngress returns the name of the ingress status with the most recent condition transition time,
diff --git a/pkg/router/controller/status_test.go b/pkg/router/controller/status_test.go
index 0d8ce8ec..73ccde5b 100644
--- a/pkg/router/controller/status_test.go
+++ b/pkg/router/controller/status_test.go
@@ -1749,7 +1749,7 @@ func Test_removeIngressCondition(t *testing.T) {
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
 			baselineIngressState := findIngressForRoute(tc.route, tc.routerName).DeepCopy()
-			changed, _, latest, original := removeIngressCondition(tc.route, tc.routerName, tc.conditionType)
+			changed, _, latest := removeIngressCondition(tc.route, tc.routerName, tc.conditionType)
 
 			// Compare expected route, but ignore LastTransitionTime since that is generated
 			cmpOpts := []cmp.Option{
```

Reply:
> I didn't really chase this through as I didn't run the tests but the diff seems small (for a future PR).

That's the thing—the fake tracker used in tests actually does need the argument.  If dropping the argument makes the tests more contorted or reduces our ability to test the tracker, then the change might not be worth it.